### PR TITLE
Fix position of new widgets when upgrading workbench versions

### LINKED ISSUE #31177: MemoryUsage Widget appears in the wrong place
Found by @thomashampson 

### Expected behavior
The new memory widget appears in the top right of the main Workbench window
### Actual behavior
If there was a previously installed version of Workbench, then a layout for the main window has been saved. When the new Memory Usage Widget is added, it takes a full column on the right, whereas we expect it to appear above the Messages box.
### Steps to reproduce the behavior
Install an older version on Mantid (release 6.0 is fine) and a recent nightly which includes the memory widget.

- Open the older version, maybe even move the different toolboxes around on the main window (e.g. drag the algorithms box to the right of the script editor)
- Successfully close the old verison (no segfault) and the layout will be saved (maybe in mantidworkbench.ini)
- Open the recent nightly. Notice how the layout you chose is preserved and that the memory widget is awkwardly added to the end. 
image (2)

- If you select View > Restore Default Layout then it gets put in the correct place.

### Platforms affected
Seen on Windows and IDAaaS, recent nightly.

### PR DESCRIPTION
**Pull request by joseph-torsney**
Workbench v6.1.0 adds a memory bar widget. If there is an existing `.ini` file from a previous version of workbench, the mainwindow state from this old version will be restored. Since the old state does not include the new widget, it gets plonked on its own in a new column (see #31177). 

This PR adds a version number to the state of the mainwindow. When restoring the state, if the version numbers mismatch, the state will not be applied and the default layout will remain. 

When applying a layout saved by the user on an old version, accessible from `View > User Layouts`, the user will be warned that workbench will attempt to restore the layout, and that it may look a bit off. If workbench is able to restore the layout, by iterating over the version numbers, the user will be informed how to remove the warning. Otherwise, the layout will not be restored.

**To test:**
1. Checkout to a commit before the memory bar widget was added to master, e.g. `git checkout e0c3d704~1`
2. Launch workbench and modify the layout.
3. Save this as a new layout in `File > Settings` 
4. Close workbench.
3. `git test-pr 31194`
4. Launch workbench, now the correct default layout should be shown. 
5. Load the old layout `View > User Layouts`. The warning should appear. 
6. Check that workbench did it's best at restoring the layout if successful.

Fixes #31177

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by thomashampson — CHANGES_REQUESTED** [review]
This does solve the immediate issue, but I think we can improve it to display a message to the user if they load a saved user layout that doesn't include the memory widget, rather than relying on them reading the release notes.

File: qt/applications/workbench/workbench/app/mainwindow.py [hunk 1]
```diff
@@ -764,3 +765,30 @@ def writeSettings(self, settings):
     def override_python_input(self):
         """Replace python input with a call to a qinputdialog"""
         builtins.input = QAppThreadCall(input_qinputdialog)
+
+    def attempt_to_restore_state(self, state):
+        success = self.restoreState(state, SAVE_STATE_VERSION)
+        if success:
```
Comment (thomashampson) [inline qt/applications/workbench/workbench/app/mainwindow.py:771]: no need for the local variable

File: qt/applications/workbench/workbench/app/mainwindow.py [hunk 2]
```diff
@@ -764,3 +765,30 @@ def writeSettings(self, settings):
...
+            "Layout Restoration",
+            "The selected layout is incompatible with this version of Workbench. Workbench will attempt to restore \\"
+            "the layout, but it may appear differently from before.\nDo you wish to continue?",
+            QMessageBox.Yes | QMessageBox.No)
+        if not reply == QMessageBox.Yes:
+            return
+
+        for version in range(0, SAVE_STATE_VERSION):
+            success = self.restoreState(state, version)
+            if success:
```
Comment (thomashampson) [inline qt/applications/workbench/workbench/app/mainwindow.py:786]: same as above, can just have:
```suggestion
            if self.restoreState(state, version):
```

**Review by Pasarus — APPROVED** [review]

**Review by thomashampson — APPROVED** [review]

**Comment by martyngigg** [comment]
Nice work everyone on catching and fixing this.

**Pull request merged**